### DISCUSSION TITLE
CsvWriter freezing

### DISCUSSION THREAD

**Discussion by petersaj** [bonsai-rx/bonsai#1534] [General]
I have a rare issue where CsvWriter appears to stop writing mid-entry, and sometimes this is accompanied by workflow issues. Unfortunately the bug is too rare to reproduce and troubleshoot, but here are some examples:

This workflow draws a sequence of visual stimuli - when this one errors, I believe the stimuli keep being drawn and the workflow ends properly, but then the CSV file ends mid-way through one of the entries: 
[lcr_passive.bonsai.txt](https://github.com/bonsai-rx/bonsai/files/12407521/lcr_passive.bonsai.txt)
![image](https://github.com/bonsai-rx/bonsai/assets/17985056/e535510d-03b9-4943-b735-606821e986ed)

This workflow has an interactive stimulus - when this one errors, the CSV file also ends mid-way through an entry, but the stimulus also becomes non-interactive (always at the start of being drawn):
[stim_wheel_right_stage2.bonsai.txt](https://github.com/bonsai-rx/bonsai/files/12407522/stim_wheel_right_stage2.bonsai.txt)
![image](https://github.com/bonsai-rx/bonsai/assets/17985056/2989f73d-30d7-423f-9ee3-378628a08715)

I saw in this issue (https://github.com/orgs/bonsai-rx/discussions/981) that a swap for reactive/shaders node was recommended - I already use the shaders one, but maybe my issues is similarly BonVision-related? Should I be trying to use BonVision's EventLogger instead of the standard CsvWriter?

**Comment by bruno-f-cruz** [comment]
Hey,

Does bonsai also crash when these bugs happen? How are you closing the workflow? A you also include the workflow file that minimally reproduces the bug?

**Comment by petersaj** [comment]
Yup the two workflows I included above (lcr_passive and stim_wheel_right_stage2) have both had the issue - added the .txt suffix to upload. Bonsai doesn't seem to crash when this happens, and the workflow either closes itself (lcr_passive) or is closed by listening to an outside stop signal (stim_wheel_right_stage2), both of which work normally when this csvwriter issue happens.

**Reply by bruno-f-cruz** [reply]
I wonder if the writer is not having enough time to flush the last values from your task logic somehow before being forced to close (since the sequence that determines the end of the observable is independent). Can you try adding a 1 second `Delay` between `Stop` and `WorkflowOutput` ? Just to force the task logic portion to finish but give time to the writer to flush everything.

**Reply by petersaj** [reply]
The `lcr_passive`  has a 2 second delay before the workflow output and `stim_wheel_right_stage2` is stopped externally (the logic gets tripped up somewhere to make the task stop working, but it still closes cleanly), so it doesn't seem like it's an issue with closing before everything's written. It does seem to always happen around the time of a stimulus onset, I'll try to keep an eye on it to catch it in the act.

**Reply by bruno-f-cruz** [reply]
That is indeed weird. If Bonsai crashes/hangs when this happens, it is unsurprising that you get a corrupted CSV file. But I don't see why this would happen if bonsai closes cleanly. Please keep us posted!

**Comment by petersaj** [comment]
I've possibly found the solution to this - it may have been the order of `MulticastSubject`.

I have two `MulticastSubject` one after another, one to write the CSV, and one to draw something (change a photodiode square). If I have the CSV one before the draw one, it sometimes messes up the CSV by stopping writing mid-entry. If I have the draw one before the CSV one, it seems consistently successful so far. 

Here's the order that seems to work, whereas the bad order was to have `PhotodiodeSquare` before `StimOn`: 

![image](https://github.com/bonsai-rx/bonsai/assets/17985056/948dcea9-efb9-45e8-b1b6-a886926278c4)

Maybe there's some kind of conflict if it's trying to write and draw at the same time? Guess it makes sense to put the CSV writing at the end of a series of things.
